Fix Randomizer.NextDecimal bias

jnm2 opened the pull request:
Fixes #3228

CharliePoole commented on src/NUnitFramework/framework/Internal/Randomizer.cs:
```diff
@@ -563,18 +563,67 @@ public string GetString()
         /// <summary>
         /// Returns a random decimal between positive zero and the specified maximum.
         /// </summary>
         public decimal NextDecimal(decimal max)
```
I think we need to document the algorithm in a comment.

jnm2 replied:
Good idea.

CharliePoole commented on src/NUnitFramework/framework/Internal/Randomizer.cs:
```diff
@@ -563,18 +563,67 @@ public string GetString()
         /// </summary>
         public decimal NextDecimal(decimal max)
         {
-            return NextDecimal() % max;
```
I'm almost ready to bet that we have other uses of modulo with other numeric types. Do we need to test them too?

jnm2 replied:
Yes. Should I open a separate issue for non-decimal types rather than do everything in this PR?

CharliePoole replied:
I'm for small PRs!

jnm2 replied:
I just checked and I don't see anything else in Randomizer.cs that is using modulo without also correcting for its bias.

jnm2 commented:
> I think this may be the first test we have written of the distribution of random values. Do we need others? Should they be separate from our general unit tests?

It is. I'm also realizing it could be a lot more thorough; each bit of the same 10,000 decimals can be separately checked for flat distribution. Also, I should test 96 maximums, one for each possible number of significant bits.

Because we were using `Next(0, int.MaxValue)`, the `low`, `mid`, and `high` ints used to create the decimals never had the high bit set. That meant `NextDecimal()` never generated a number between `decimal.MaxValue / 2` and `decimal.MaxValue` as well as having some gaps.

Because the tests run so fast (10 ms on my machine), I'm not thinking of a benefit to keeping distribution tests in a separate test run. Maybe a separate test class?

jnm2 commented:
I need to fix this too: 10,000 isn't enough iterations for a 0.01 precision.

It isn't true randomness if there isn't *some* chance that the distribution of a particular random sequence is biased, so test failure is always possible with these tests. It's just a matter of calculating the likelihood and making it sufficiently small in the same way that guid collisions are sufficiently negligible.

Maybe the test could keep running up to a time limit before failing, just as long as there is a minimum number of iterations.

CharliePoole commented:
I wasn't thinking of efficiency so much as the fact that this is a bit out of our wheelhouse. We're more users of pseudo-random number generation than creators, even though some of us may have a degree of experience with them.

Ideally, we would be using somebody else's library and testing library. Ideally, we 
might allow users to replace the underlying sequence generator.

So for me, separation implies a possible future division of labor.

But for now, a separate class is fine. :smile:

jnm2 commented:
It seems to me that by synthesizing our own NextDecimal methods at all, we're placing ourselves firmly in the pseudorandom value creators' seat. Our implementation is either good or poor and if we have it at all, I'd like it to be good.

CharliePoole commented:
Agreed. I'm speculating about whether we can get out of that position. I do know there are some very good generalized suites for testing random number generators. I also know that our synthesized values can never get better than the underlying generator, which adds a bit of complexity.

jnm2 commented:
@CharliePoole Ready for review again.

CharliePoole commented:
Seems like I have to start over from the beginning because of the force push.

jnm2 commented:
If you click this link above, it shows you the diff:

![image](https://user-images.githubusercontent.com/8040367/58052922-17368980-7b24-11e9-9f41-03aec19e6e8c.png)

jnm2 commented:
That's not as helpful as I thought since you can't go commit by commit. I'm sorry.

jnm2 commented:
The original review is also not lost. I clicked Resolve to collapse it:

![image](https://user-images.githubusercontent.com/8040367/58053045-6ed4f500-7b24-11e9-93e3-f0ad553d67f5.png)

CharliePoole commented:
GitHub says that @jnm2 dismissed my stale review. Not sure but I think that means the forced push. It's useful to be able to look back at what one reviewed and asked for as new requests for review arrive. We should try to avoid situations where re-review is required and the reviewer can no longer access the old review.

jnm2 commented:
I'll prefer to be forward-only in the future to make reviewing easier. I think I will still track which things I have finished by collapsing with the Resolve button though.

It's not possible for review comments to be lost. Any new push, even a non-force push, will dismiss all reviews as stale. Dismissing a review only clears the status. It doesn't remove any information that had been added to the thread.

CharliePoole commented:
As a reviewer, I want to click on my review and see what I wrote. If GitHub says it can't find the information, then my review is "lost" to me even if there is some further search or steps I don't know about that would allow me to find them.

Part of this problem, of course, is in the policy that requires a re-review for every change that is made. That's because we (1) have a strict policy of two committers to approve a review and (2) have decided to enforce that policy through GitHub rather than socially.

jnm2 commented:
GitHub doesn't say that it can't find the information since that's not what 'dismiss' means and there is UI in the same location as the original information saying "click to show."

Both of these effects (dismissing reviews on non-force pushes and collapsing addressed comments) would still be in effect even if all the GitHub requirements were turned off, I think.

CharliePoole commented:
Trust me... that __is__ what GitHub UI said when I clicked. :smile: 

The previous review is gone - either really gone or gone for practical purposes, since I can't find it. I figure if someone with my experience can't immediately find the information, then it's gone for most people. This isn't a criticism, by the way, of anything you did and I hope it did not sound that way.

jnm2 commented:
That's a mystery. Earlier on desktop and now on my phone I can see everything that has been said in the previous review.

Thanks! I don't take it as a criticism. I'm commenting to clarify that I don't see any action the NUnit team could take to avoid these outcomes (besides report them as bugs).

oznetmaster commented:
For some reason I cannot determine, on my CF build, the NextDecimalNotBiased test fails for values 0, 32, and 64.  All other values work correctly.

For those three values, the result 0.0d.

Any thoughts as to where to look?

The NextDecimalWithMaximumNotBiased test passes as well.

CharliePoole commented:
@jnm2 Are you referring to the code comments? I can see those. It's the review itself I can't find. For example, did I previously approve it or ask for changes? What reasons did I give?

IMO, the action that could be taken would be to not rely on GitHub software to enforce our cooperative agreements. Then no bug reports or fixes would be needed. However, this is probably better not discussed here.

jnm2 commented:
That means that `Random.Next(int.MinValue, int.MaxValue)` is returning numbers whose least significant bit is never set.

For CF then I'd recommend RawInt32 to be implemented using NextBytes and BitConverter.ToInt32. NextBytes takes a 1 random sample per byte on .NET Framework, whereas `Next(int.MinValue, int.MaxValue)` takes two samples.

You would find the sourceof.net implementation interesting for `Next(int, int)`. It has a special case that I'm guessing CF does not have. I'm going to assume CF takes only one sample.

jnm2 commented:
@oznetmaster Hey, where did your comment go?

oznetmaster commented:
It is still there

CharliePoole commented:
@jnm2 Yes, and that implies any tests we write have to  take into account the underlying Random distribution, which can differ according to platform and possibly even version. Wow!

jnm2 commented:
@oznetmaster I got mixed up, sorry. [Here's the netfx implementation.](https://referencesource.microsoft.com/#mscorlib/system/random.cs,ddd1e9e09c70bab5)
Does the NextBytes route seem workable?

oznetmaster commented:
I am testing it now

jnm2 commented:
@CharliePoole 

> Are you referring to the code comments? I can see those. It's the review itself I can't find. For example, did I previously approve it or ask for changes? What reasons did I give?

It is here: https://github.com/nunit/nunit/pull/3241#pullrequestreview-237151229 (the reasons you gave are right above the code comments)
You originally did a comment review rather than approving or ask for changes. I noticed that right away. If you had requested changes, that line would still be showing a red +/- icon and would say "CharliePoole requested changes" rather than "CharliePoole left a comment."

oznetmaster commented:
@jnm2 That fixed the problem.

Does making that change standard make the code more platform independent in the long run?

RawUInt32 and RawUInt64 both use that code, so why not RawInt32?

CharliePoole commented:
@jnm2 So I think you are saying that when I do a review and check "Comment"
it's not recorded as a review at all. That would explain why I couldn't
find it for sure. What I was looking for was something exactly like the box
in which I entered the review in the first place, because (as I've often
said) I implement a stateless protocol with respect to comments and
reviews. Each time I look I'm starting over because I comment on too many
things for too many people to remember state for each one. :smile:

On Tue, May 21, 2019 at 12:40 AM Joseph Musser <notifications@github.com>
wrote:

> @CharliePoole <https://github.com/CharliePoole>
>
> Are you referring to the code comments? I can see those. It's the review
> itself I can't find. For example, did I previously approve it or ask for
> changes? What reasons did I give?
>
> It is here: #3241 (review)
> <https://github.com/nunit/nunit/pull/3241#pullrequestreview-237151229>
> (the reasons you gave are right above the code comments)
> You originally did a comment review rather than approving or ask for
> changes. I noticed that right away. If you had requested changes, that line
> would still be showing a red +/- icon and would say "CharliePoole requested
> changes" rather than "CharliePoole left a comment."
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nunit/nunit/pull/3241?email_source=notifications&email_token=AAUN7BMRLOYUSG3SIC5V62DPWMSGTA5CNFSM4HMU7XTKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV2IJIA#issuecomment-494175392>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUN7BNSNVOPYTJ2MT6YVX3PWMSGTANCNFSM4HMU7XTA>
> .
>

jnm2 commented:
@CharliePoole It is recorded just like the other two kinds in all the same ways, including in the summary at the bottom of the page before the PR is merged. It just uses a different icon (one of three) and a different phrase (one of three) underneath the text "CharliePoole reviewed 7 days ago."

jnm2 commented:
@oznetmaster I'd ordinarily not support it because it's a bug if a framework only returns even integers when you ask for a range. However, it just struck me that `Next(int.MinValue, int.MaxValue)` excludes the number `int.MaxValue`! It's almost impossible to detect this in a test (chance is one in four billion) but it will force me to switch to NextBytes as the next most obvious implementation.

PR incoming!

oznetmaster commented:
@jnm2 The CF NUnit framework is not only returning even integers if a range is specified.  All 96 tests are being run correctly.  It was only the 0, 32, 64 tests that were failing.

oznetmaster commented:
As a point of information, I just ran the RandomizerBiasTests (fixed as above) on the CF framework 100 times, and it passed every time.

jnm2 commented:
What do you mean by the 0, 32, and 64 tests failing? That would be three of the 96 tests, corresponding to bits 0, 32, and 64, wouldn't it?

oznetmaster commented:
Yes.  Test cases 0, 32, and 64 (generated by the Range attribute).

Once the changes were made (as above) all the test cases pass all the time.

jnm2 commented:
Right, so the only way that could be is if CF's `Next(int.MinValue, int.MaxValue)` fails to set the least significant bit (ones place). Does that make sense?

oznetmaster commented:
CF's Next:

```
public virtual int Next(int minValue, int maxValue)
{
    if (minValue > maxValue)
    {
        throw new ArgumentOutOfRangeException();
    }
    int num = maxValue - minValue;
    if (num >= 0)
    {
        return (((int) (this.Sample() * num)) + minValue);
    }
    long num2 = maxValue - minValue;
    return (((int) ((long) (this.Sample() * num2))) + minValue);
}
```

jnm2 commented:
> For those three values, the result 0.0d.

This means that bits 0, 32, and 64 were set 0 out of 100,000 times. Those bits came directly from bit 0 of `Next(int.MinValue, int.MaxValue)`; therefore, bit 0 of `Next(int.MinValue, int.MaxValue)` was set 0 out of 300,000 tries.

[Here's the netfx implementation](https://referencesource.microsoft.com/#mscorlib/system/random.cs,ddd1e9e09c70bab5) I linked above:

```cs
      public virtual int Next(int minValue, int maxValue) {
          if (minValue>maxValue) {
              throw new ArgumentOutOfRangeException("minValue",Environment.GetResourceString("Argument_MinMaxValue", "minValue", "maxValue"));
          }
          Contract.EndContractBlock();
          
          long range = (long)maxValue-minValue;
          if( range <= (long)Int32.MaxValue) {  
              return ((int)(Sample() * range) + minValue);
          }          
          else { 
              return (int)((long)(GetSampleForLargeRange() * range) + minValue);
          }
      }

      private double GetSampleForLargeRange() {
          // The distribution of double value returned by Sample 
          // is not distributed well enough for a large range.
          // If we use Sample for a range [Int32.MinValue..Int32.MaxValue)
          // We will end up getting even numbers only.
 
          int result = InternalSample();
          // Note we can't use addition here. The distribution will be bad if we do that.
          bool negative = (InternalSample()%2 == 0) ? true : false;  // decide the sign based on second sample
          if( negative) {
              result = -result;
          }
          double d = result;
          d += (Int32.MaxValue - 1); // get a number in range [0 .. 2 * Int32MaxValue - 1)
          d /= 2*(uint)Int32.MaxValue - 1  ;              
          return d;
      }
```

This is why I'm so sure that CF is faulty, but you can remove all doubt by testing whether this ever completes:

```cs
var rng = new Random();

while ((rng.Next(int.MinValue, int.MaxValue) & 1) == 0)
{
}
```

When I run the above code on .NET Framework but using the CF implementation you posted, it never completes.
